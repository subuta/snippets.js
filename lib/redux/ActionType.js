import { build, format, snippets as s } from 'bld.js'
import _ from 'lodash'
import pluralize from 'pluralize'

const toActionType = (prefix, type, suffix = '') => {
  if (!type) return _.toUpper(_.snakeCase(prefix))
  return _.toUpper(_.snakeCase(`${prefix}_${type}${suffix && '_' + suffix}`))
}

export const ActionTypeDef = function (...args) {
  const ACTION_TYPE = toActionType.apply(this, args)
  return build`
    const ${ACTION_TYPE} = ${s.stringify(ACTION_TYPE)}
  `
}

export default {
  raw: toActionType,
  set: (model) => toActionType('set', pluralize(model)),
  setIds: (model) => toActionType('set', pluralize.singular(model), 'ids'),
  request: (model) => toActionType('request', pluralize(model)),
  failure: (model) => toActionType('request', pluralize(model), 'failure'),
  create: (model) => toActionType('create', pluralize.singular(model)),
  show: (model) => toActionType('show', pluralize.singular(model)),
  all: (model) => toActionType('index', pluralize(model)),
  update: (model) => toActionType('update', pluralize.singular(model)),
  destroy: (model) => toActionType('delete', pluralize.singular(model)),
}
