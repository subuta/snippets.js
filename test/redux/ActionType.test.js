import test from 'ava'
import { build, format, snippets as s } from 'bld.js'

import ActionType, { ActionTypeDef } from 'lib/redux/ActionType'

test.beforeEach(async (t) => {

})

test.afterEach((t) => {
})

test('should create redux ActionType with single argument', async (t) => {
  const code = build`
    ${ActionType.raw('setBooks')}
  `

  const expected = build`
    SET_BOOKS
  `

  t.is(format(code), format(expected))
})

test('should create redux ActionType with prefix and type', async (t) => {
  const code = build`
    ${ActionType.raw('set', 'books')}
  `

  const expected = build`
    SET_BOOKS
  `

  t.is(format(code), format(expected))
})

test('should create redux ActionType for setter', async (t) => {
  const code = build`
    ${ActionType.set('books')}
  `

  const expected = build`
    SET_BOOKS
  `

  t.is(format(code), format(expected))
})

test('should create redux ActionType for request', async (t) => {
  const code = build`
    ${ActionType.request('books')}
  `

  const expected = build`
    REQUEST_BOOKS
  `

  t.is(format(code), format(expected))
})

test('should create redux ActionType for failure', async (t) => {
  const code = build`
    ${ActionType.failure('books')}
  `

  const expected = build`
    REQUEST_BOOKS_FAILURE
  `

  t.is(format(code), format(expected))
})

test('should create redux ActionType Def', async (t) => {
  const code = build`
    ${ActionTypeDef('setBooks')}
  `

  const expected = build`
    const SET_BOOKS = 'SET_BOOKS'
  `

  t.is(format(code), format(expected))
})