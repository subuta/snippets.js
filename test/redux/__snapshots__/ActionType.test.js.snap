// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux/ActionType should create redux ActionType Def 1`] = `
"const SET_BOOKS = 'SET_BOOKS'
"
`;

exports[`redux/ActionType should create redux ActionType for all(index) Action 1`] = `
"INDEX_BOOKS
"
`;

exports[`redux/ActionType should create redux ActionType for create Action 1`] = `
"CREATE_BOOK
"
`;

exports[`redux/ActionType should create redux ActionType for destroy(delete) Action 1`] = `
"DELETE_BOOK
"
`;

exports[`redux/ActionType should create redux ActionType for failure 1`] = `
"REQUEST_BOOKS_FAILURE
"
`;

exports[`redux/ActionType should create redux ActionType for request 1`] = `
"REQUEST_BOOKS
"
`;

exports[`redux/ActionType should create redux ActionType for setter 1`] = `
"SET_BOOKS
"
`;

exports[`redux/ActionType should create redux ActionType for show Action 1`] = `
"SHOW_BOOK
"
`;

exports[`redux/ActionType should create redux ActionType for update Action 1`] = `
"UPDATE_BOOK
"
`;

exports[`redux/ActionType should create redux ActionType with prefix and type 1`] = `
"SET_BOOKS
"
`;

exports[`redux/ActionType should create redux ActionType with single argument 1`] = `
"SET_BOOKS
"
`;
